lib/content-server.ts

```diff
@@ -43,95 +82,79 @@ export class ContentServerStack extends cdk.NestedStack {
+            logGroupName: `/opentuna/${usage}/cloudwatch-agent`,
+            removalPolicy: cdk.RemovalPolicy.DESTROY
+        });
+        const cloudWatchAgent = taskDefinition.addContainer("cloudwatch-agent", {
+            image: ecs.ContainerImage.fromRegistry('amazon/cloudwatch-agent:latest'),
+            logging: new ecs.AwsLogDriver({
+                streamPrefix: usage,
+                logGroup: cloudWatchAgentlogGroup,
             }),
+            essential: false,
+            secrets: {
+                "CW_CONFIG_CONTENT": ecs.Secret.fromSsmParameter(param),
+            }
         });
 
         const service = new ecs.FargateService(this, `${usage}Fargate`, {
             cluster: props.ecsCluster,
-            desiredCount: 2,
+            desiredCount: 1,
             platformVersion: ecs.FargatePlatformVersion.VERSION1_4,
             taskDefinition,
         });
-        service.taskDefinition.executionRole?.grantPassRole(customTaskDefinition.grantPrincipal);
-        service.taskDefinition.taskRole.grantPassRole(customTaskDefinition.grantPrincipal);
 
-        // we will replace this hack when CDK supports EFS in Fargate
-        (service.node.tryFindChild('Service') as ecs.CfnService)?.addPropertyOverride(
-            'TaskDefinition',
-            customTaskDefinition.getResponseField('taskDefinition.taskDefinitionArn'),
-        );
+        // setup cloudwatch agent permissions
+        // allow execution role to read ssm parameter
+        param.grantRead(service.taskDefinition.executionRole!);
+        service.taskDefinition.executionRole!.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy'));
+        service.taskDefinition.taskRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy'));
 
-        props.listener.addTargets('ContentServer', {
+        const targetGroup = props.listener.addTargets('ContentServer', {
             port: httpPort,
             protocol: elbv2.ApplicationProtocol.HTTP,
-            targets: [ service ],
+            targets: [service],
             healthCheck: {
                 enabled: true,
                 timeout: cdk.Duration.seconds(15),
             },
         });
 
+        // auto scaling
+        const scaling = service.autoScaleTaskCount({
+            minCapacity: 1,
+            maxCapacity: 16
+        });
+        const bytesSentEth1 = new cloudwatch.Metric({
+            namespace: 'OpenTuna',
+            metricName: 'net_bytes_sent',
+            dimensions: {
+                interface: "eth1"
+            },
+            statistic: cloudwatch.Statistic.AVERAGE,
+        });
+        // each task instance has about 500Mbps bandwidth
+        // 500Mbps * 60s/min / 8b/B = 3750MB/min
+        // must use direct metric because of api limitation
+        // so only eth1 is used
+        scaling.scaleToTrackCustomMetric('NetworkBandwidthScaling', {
```

Comment:
Per doc https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-target-tracking.html, the auto scaling support more than one policy.

I prefer to add another policy to more smoothly scale in,

1. set the metric of 50% of maximum traffic
1. set 10m as cool down period of scale in
1. set a super large period of cool down of scale out

Those two policies will make the scaling out when the metrics exceeding 3GB metrics threshold. And the scaling in must the traffic is lower than 50% utilities.